undofx/src/main/java/org/fxmisc/undo/impl/UndoManagerImpl.java

```diff
@@ -209,7 +229,7 @@ private void addChange(C change) {
     @SuppressWarnings("unchecked")
     private C[] merge(C c1, C c2) {
         Optional<C> merged = merge.apply(c1, c2);
-        if(merged.isPresent()) {
+        if(merged.isPresent() && isNotIdentity.test(merged.get())) {
             return (C[]) new Object[] { merged.get() };
         } else {
             return (C[]) new Object[] { c1, c2 };
```

Comment:
> Yes, that's what I intended.
> 
> Taking RichTextFX as an example, say you type in hello and then hit backspace 5 times. With my version, the undo history will now be empty. With yours, it will contain 2 items:
> 
>     at 0, insert `h`
>     at 0, delete `h`
>
> I would say yours is as far from what actually happened as mine.
> 
> Plus, why merge the first 4 deletions and not the 5th one? Seems somewhat irregular and arbitrary.
>
> Plus, if the user didn't want changes to annihilate, they could have provided merge that would return Optional.empty() instead of an identity change. Thus this behavior could be implemented without any new support from UndoFX.

Took me a while to understand what you meant. Rephrasing it in my own words, if a user typed in "hello", this would result in five changes being emitted by the change stream. When `merge` is called, these are merged into one change `insert: "hello" | removed: ""`. When the user deletes the last four letters, the `merge` will gradually remove the letters in the `insert` part of the change: `hello -> hell -> hel -> he -> h`. On the final deletion, it would attempt to merge `insert: "h" | removed: ""` with `insert: "" | removed: "h"`. The result would be an identity change `inserted: "" | removed: ""`. 

There are two ways to handle this situation. The first is what my code does: "the merge was successful, but it's an identity change, so just re-store the previous change and store the new pushed change". As a result, the queue would have two changes. The second is what you have said should occur: "the merge was successful, but it's an identity change, so act as though neither the stored change nor the pushed change were ever done in the first place." In effect, the changes have been "annihilated." 
If my approach is used and one doesn't want "change annihilation" in their code, there is nothing they can do about it. If your approach is used and they don't want it, they can simply modify the `merge` to return `Optional.empty()` in the situation where a successful merge would return an identity change.
Thus, your approach is more flexible between the two.